rollup.config.js

```diff
@@ -37,44 +39,89 @@ const resolveOpts = {
   extensions: [".ts", ".js"],
 };
 
-export default (async () => [
-  {
-    input: `./src/single-spa${isProduction ? "" : ".profile"}.ts`,
-    output: [
-      {
-        file: `./lib/umd/single-spa${isProduction ? ".min" : ".dev"}.cjs`,
-        format: "umd",
-        name: "singleSpa",
-        sourcemap: true,
-        banner: generateBanner("UMD"),
-      },
-      {
-        file: `./lib/system/single-spa${isProduction ? ".min" : ".dev"}.js`,
-        format: "system",
-        sourcemap: true,
-        banner: generateBanner("SystemJS"),
-      },
-      {
-        file: `./lib/esm/single-spa${isProduction ? ".min" : ".dev"}.js`,
-        format: "esm",
-        sourcemap: true,
-        banner: generateBanner("ESM"),
-      },
-    ],
-    plugins: [
-      replace(replaceOpts),
-      resolve(resolveOpts),
-      babel(babelOpts),
-      commonjs(),
-      isProduction &&
-        (await import("@rollup/plugin-terser")).default(terserOpts),
-      useAnalyzer && analyzer(),
-    ],
-  },
-])();
-
 function generateBanner(format) {
   return `/* single-spa@${packageJson.version} - ${format} - ${
     isProduction ? "prod" : "dev"
   } */`;
 }
+
+const commonPlugins = [
+  replace(replaceOpts),
+  resolve(resolveOpts),
+  babel(babelOpts),
+  commonjs(),
+  isProduction && (await import("@rollup/plugin-terser")).default(terserOpts),
+  useAnalyzer && analyzer(),
+];
+
+const commonConfiguration = {
+  input: `./src/single-spa${isProduction ? "" : ".profile"}.ts`,
+};
+
+export default async () => [
+  // UMD build
+  {
+    ...commonConfiguration,
+    output: {
+      file: `./lib/umd/single-spa${isProduction ? ".min" : ".dev"}.cjs`,
+      format: "umd",
+      name: "singleSpa",
+      sourcemap: true,
+      banner: generateBanner("UMD"),
+    },
+    plugins: [
+      ...commonPlugins,
+      typescript({
+        declarationDir: "./lib/umd/types-tsc",
+        outDir: "./lib/umd",
+      }),
+    ],
+  },
+  // ESM build
+  {
+    ...commonConfiguration,
+    output: {
+      file: `./lib/esm/single-spa${isProduction ? ".min" : ".dev"}.js`,
+      format: "esm",
+      sourcemap: true,
+      banner: generateBanner("ESM"),
+    },
+    plugins: [
+      ...commonPlugins,
+      typescript({
+        declarationDir: "./lib/esm/types-tsc",
+        outDir: "./lib/esm",
+      }),
+    ],
+  },
+  // System build
+  {
+    ...commonConfiguration,
+    output: {
+      file: `./lib/system/single-spa${isProduction ? ".min" : ".dev"}.cjs`,
+      format: "system",
+      sourcemap: true,
+      banner: generateBanner("SystemJS"),
+    },
+    plugins: [...commonPlugins],
+  },
+  // For compatibility with node16 module resolution, types are bundled
+  // Bundle ESM types
+  {
+    input: `./lib/esm/types-tsc/single-spa.d.ts`,
+    output: {
+      file: `./lib/esm/types/single-spa.d.ts`,
```

Comment:
Since the types are bundled into one file, I don't see why the `types` subfolder is needed

```suggestion
      file: `./lib/esm/single-spa.d.ts`,
```